Debugger: add top-level crate and async wrapper allowing for a "debugger environment".

cfallin opened the pull request:
The debug support in Wasmtime so far is structured around async callbacks that occur at certain kinds of events, like breakpoints. This is a suitable foundation but makes for an awkward implementation of a top-level debugger implementation, which likely has an event loop dealing with user commands (via a UI or a protocol connection) and expects to perform actions such as "run until next breakpoint".

This PR introduces a new crate that wraps a `Store` in a `Debugger`. This wrapper embodies an inner async body that can perform whatever actions it likes on the `Store` that is passed back in. This inner body is spawned as an async task. The debugger wrapper registers its own `DebugHandler` callback that communicates with the outside world via bidirectional command/response queues.

On the "outside", the `Debugger` presents an interface suitable for inserting into a debug protocol server or UI: an async method that runs until next event and returns that event, and a method that permits querying or modifying the store whenever the `run` method is not executing. The latter operates by sending a closure over the queue, because the `Store` must continue to be owned by the async task that is (still) running and suspended in async callbacks.

Right now, this is exercised only via a few unit tests, but the intent is to next build up the "top half" of the debugger using this abstraction, e.g. by running a gdbstub protocol server (likely as a Wasm component in a "debug-main WIT world" -- RFC needed for this).

Also, when we eventually move debugging over to native use of `run_concurrent`, this paradigm should remain mostly unchanged at this level of API: there can still be an object that has an async method that runs and yields the next event, and there can still be a method that takes a closure that can operate (within its scope only) on the `Store`.

A few warts that I could use feedback on:

- Cancelation safety is weird. Fibers panic when dropped before execution of their body completes, and this seems to mean that we can't allow a `Debugger` to drop early (or at least, the `tokio::test` unit test that owns the runtime that runs the async task to finish before the debugged body completes!). If there is a better way to handle cancelation safety here, I'm all ears.

- It's not clear to me if the boxed-closure-and-`Any` approach to providing access to the `Store` is the best we can do, but I suspect it is.

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

fitzgen commented on crates/debugger/src/lib.rs:
```diff
@@ -0,0 +1,636 @@
+    /// callback (or in the initial trampoline waiting for the first
+    /// `Continue`). `Response::Stopped` has been received. Outer
+    /// logic has not sent any commands.
+    Stopped,
```
Maybe name this `Paused`, because the name is a bit ambiguous with `Complete` otherwise (a program stops when it is finished, does it not?)

fitzgen commented on crates/debugger/src/lib.rs:
```diff
@@ -0,0 +1,636 @@
+/// from the `JoinHandle` above, are what allow us to implement a
+/// cancel-safe version of `Debugger::run` below.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum DebuggerState {
```
Could be good to have an ascii diagram of the state machine here. Assuming we start in the `Paused` state, effectively always beginning at a breakpoint just before the program's first instantiation, I'm thinking something like

```text
  |
  |
  V
Paused <---------> Queried
  ^
  |                  
  |
  V
Running
  |
  |
  V
Complete
```

But now that I think about it some more, it seems like complete and paused are actually the same for "reactor" style programs? That is, those that are a bag of exported functions that can be called in any order, rather than those defining just a `main`. This is because you start out in the initial paused state, call into some exported function and do stuff and then that can complete but you can always call back in again and do more stuff and debug these subsequent entries into the program as well.

This also makes me think of something that I don't think we've discussed thus far in our debugging conversations: debugging the start function and instantiation in general. For example, what if a data segment's initialization traps due to out-of-bounds memory access? We want to support debugging that, but I don't think we raise a debugging trap event for that yet, do we? (Obviously, we don't need to address all that in this PR, but I think this state machine should reflect the actual states we want to support in the fullness of time, IMO). So given that I think we actually want something like

```
   |
   V
Uninstantiated <---.
   ^               |
   |               | (debugger says terminate)
   |               |
   V               |
Running            |
   ^               |
   |               | 
   |               |
   V               |
Paused-------------'
   ^
   |
   |
   V
Queried
```

where uninstantiated to running involves instantiation, and returning from `main` or any other export that we called into as the entry point puts us back at paused (since we can inspect memories and such in this state, but can't in the uninstantiated state)

fitzgen reviewed (commented):
Have to context switch, but want to get my comments thus far posted.

alexcrichton commented on crates/debugger/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+[dev-dependencies]
+# Depend on `wasmtime` again to get `cranelift` so we can write unit
+# tests that build modules.
+wasmtime = { workspace = true, features = ["debug", "std", "async", "cranelift", "wat"] }
```
You can drop debug/std/async here since that'll come in through the normal dependency

alexcrichton commented on crates/debugger/src/lib.rs:
```diff
@@ -0,0 +1,636 @@
+        mut store: StoreContextMut<'_, T>,
+        event: DebugEvent<'_>,
+    ) -> impl Future<Output = ()> + Send {
+        async move {
```
This should be define-able with `async fn handle(...)` and you don't need to worry about `Send`

cfallin commented:
> This is definitely a bug in Wasmtime that we need to fix. Can you add a failing test or gist one that I can help poke around?

Sure thing -- with [this](https://github.com/cfallin/wasmtime/commit/2d22450579d820257317fe125b634ab360291e51) commit and

```
cargo test -p wasmtime-internal-debugger -- basic_debugger
```

I see

```
thread 'test::basic_debugger' (1129985) panicked at /home/cfallin/work/wasmtime/crates/wasmtime/src/runtime/fiber.rs:199:70:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

cfallin commented:
I'm not sure I understand the failure now -- I added `wasmtime-publish` and it shows as an owner; I removed myself as an owner; it's still failing. @alexcrichton any ideas?

cfallin commented:
Thanks!
